torchgeo/models/farseg.py

```diff
@@ -0,0 +1,277 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+"""Foreground-Aware Relation Network (FarSeg) implementations."""
+
+import math
+from collections import OrderedDict
+from typing import List, cast
+
+import torch.nn.functional as F
+from torch import Tensor
+from torch.nn.modules import (
+    BatchNorm2d,
+    Conv2d,
+    Identity,
+    Module,
+    ModuleList,
+    ReLU,
+    Sequential,
+    Sigmoid,
+    UpsamplingBilinear2d,
+)
+from torchvision.models import resnet
+from torchvision.ops import FeaturePyramidNetwork as FPN
+
+# https://github.com/pytorch/pytorch/issues/60979
+# https://github.com/pytorch/pytorch/pull/61045
+Module.__module__ = "torch.nn"
+ModuleList.__module__ = "nn.ModuleList"
+Sequential.__module__ = "nn.Sequential"
+Conv2d.__module__ = "nn.Conv2d"
+BatchNorm2d.__module__ = "nn.BatchNorm2d"
+ReLU.__module__ = "nn.ReLU"
+UpsamplingBilinear2d.__module__ = "nn.UpsamplingBilinear2d"
+Sigmoid.__module__ = "nn.Sigmoid"
+Identity.__module__ = "nn.Identity"
+
+
+class FarSeg(Module):
+    """Foreground-Aware Relation Network (FarSeg).
+
+    If you use this model in your research, please cite 'Foreground-Aware Relation
+    Network for Geospatial Object Segmentation', Zheng et al. (2020):
+
+    * https://arxiv.org/pdf/2011.09766.pdf
+    """
+
+    def __init__(
+        self,
+        in_channels: int = 3,
+        backbone: str = "resnet50",
+        classes: int = 16,
+        backbone_pretrained: bool = True,
+    ) -> None:
+        """Initialize a new FarSeg model.
+
+        Args:
+            in_channels: number of channels per input image
+                (default=3 for RGB image)
+            backbone: name of ResNet backbone
+                (default='resnet50')
+            classes: number of output segmentation classes
+                (default=16 for multi-class segmentation)
+            backbone_pretrained: whether to use pretrained weight for backbone
+                (default=True)
+        """
+        super(FarSeg, self).__init__()  # type: ignore[no-untyped-call]
+        if in_channels != 3:
```

Comment:
`in_channels` has to be 3 for now because the resnet backbones don't naively have an option to take other sized inputs. (note: we can change this by overwriting model.conv1 as I do in the So2Sat trainer)

Reply:
In that case should we not have an `in_channels` option at all if it doesn't do anything?